scrape/scrape.go

```diff
@@ -1568,6 +1573,8 @@ loop:
 
 		// Increment added even if there's an error so we correctly report the
 		// number of samples remaining after relabeling.
+		// We still report duplicated samples here since this number should be the exact number
+		// of time series exposed on a scrape after relabelling.
```

Comment:
I don't see how this code is reached if `err` is set on line 1545; this will cause `break loop` on line 1560.

Reply:
Line 1545 sets err to `ErrDuplicateSampleForTimestamp`.
Then on line 1555 `checkAddError(..., err, ...)` is called.
If you lookup the source of `checkAddError()` you'll see that it takes an error and returns a different error depending on the input error, and with `ErrDuplicateSampleForTimestamp` it returns a `nil`:
https://github.com/prymitive/prometheus/blob/ea3ad744a9e117dbc54de5d10deeb22f8e58fbfc/scrape/scrape.go#L1758

Reply:
Thanks for that explanation.  Maybe you meant to link to https://github.com/prymitive/prometheus/blob/ea3ad744a9e117dbc54de5d10deeb22f8e58fbfc/scrape/scrape.go#L1789 ?

This increases my feeling the code should be refactored. 

But, I realise now you weren't aiming to flag an error, except through the metric.  You said this in the description.

Reply:
I moved staleness tracking calls out of `checkAddError()`, I agree that it doesn't belong there.